### ISSUE TITLE
Symbol collision leading to error in generated code

### ISSUE THREAD

**Issue by bmaclach** [JacquesCarette/Drasil#1137]
One of the errors in the generated code for glassBR happens because there is a variable `k` and also a constant `k` with a value of `2.86e-53`. This symbol collision leads to the value of the constant being used instead of the variable `k`. The generated code should be:
`y_2 = func_lin_interp(x_z_2.get(k_2), y_z_2.get(k_2), x_z_2.get(k_2 + 1), y_z_2.get(k_2 + 1), x);`
But it currently is:
```
y_2 = func_lin_interp(x_z_2.get(2.86e-53), y_z_2.get(2.86e-53), x_z_2.get(2.86e-53 + 1), y_z_2.get(2.86e-53 + 1), x);
```
In the past, we used the symbol k_2 for the variable to avoid this error, however this was undone in commit 7684b53535ba0a253299597ce7d9e985ad278030. The symbol for the variable was changed to `k` and a comment about  how `k` breaks things was removed (see line 70-71 in red and line 78 in green). I just want to make sure the change of the symbol in 7684b53535ba0a253299597ce7d9e985ad278030 was accidental and that it is okay for me to change it back to fix the code. The comment that was removed said that it should not be a problem after August 2018, so I guess the other potential issue is that Drasil should be handling symbol collisions like this one correctly, but it is not.

**Comment by JacquesCarette** [comment]
The break was definitely accidental. The collision definitely should have been detected, but clearly was not, which is a bit of a pain.  Yes, go ahead and change it back to fix the code.  I guess this issue should still be left open though, to deal with the collision issue?

**Comment by JacquesCarette** [comment]
@bmaclach could you revisit this?  What I mean is, to better understand the real source of the error. It is certainly possible to write "bad" code in Drasil+GOOL, but many of the straightforward errors should be caught before the code is actually generated. It feels to me like this is not happening in this case. I would like to know why - so that we then hopefully assign it to a summer student to fix, if it's not too big a job.

**Comment by bmaclach** [comment]
We now use UIDs when inlining constants, so the specific problem shown in the original issue description should never happen, even if the same symbol is used for multiple things. But there are other places where a symbol collision still causes problem (specifically, the `variable` function in the generator checks whether the variable is an input or constant by looking at its name [which comes from its symbol], instead of its UID).

I guess the question is what is the intended behaviour? Should we 
1. allow using the same symbol for multiple quantities, and trust the user not to do so for two quantities in the same scope, which would lead to ambiguity? Or 
2. not allow the same symbol to be used for two different things at all?

If 1, the fix is probably just a matter of updating `variable` to look at UIDs to determine whether a variable is an input or constant.
If 2, the fix would be to implement a pass on all of the symbols used in each case study, and complain if there are duplicates.

**Comment by JacquesCarette** [comment]
What do you mean by "symbol" here? If you mean what a particular concept renders to, then we most definitely _want_ to have different quantities share the same symbol. In common use, this happens all the time.

Of course, given a particular context-of-use, we can't allow that.

We don't want to trust users to not make such mistakes. We should check that.  Once we know a "context of use" (a code block), that all symbols render uniquely.

So it is like 2, but perhaps not 'globally'. So there is still work to do (not by you), but less than before.

**Comment by bmaclach** [comment]
Your interpretation of my use of "symbol" is correct: it's the rendering of the concept.

**Comment by bmaclach** [comment]
(Also note that it is probably still a good idea, if not necessary, to do the fix suggested for 1 as well)

**Comment by smiths** [comment]
I agree with the analysis above.  Globally, we have no choice but to allow different quantities to share the same symbol.  As our knowledge base expands, it will be impossible to maintain uniqueness, especially if we want our symbols to be understandable.  

As Drasil is implemented right now, is it possible for two documents to use the same concept, but render it to a different representations?  I don't know the design as well as I should, but I have the feeling that the rendering of a symbol is set globally.  If that is the case, we'll need to add more flexibility in the future.  Some authors might want to represent density by d, others might want to represent it by rho.  Given that we need to be aware of clashes, we are going to need this kind of flexibility.

I agree with @JacquesCarette that this is outside the scope of the current work by @bmaclach.

**Comment by JacquesCarette** [comment]
Right now, the rendering of a symbol is global (though split into 2, once for specifications, once for code).  We might want to have a way to import a symbol from a global database and then locally override is display representation.

**Comment by smiths** [comment]
Thanks @JacquesCarette.  Good to know that we have the local override option for the future.  As we add more examples, I'm sure the need for this will arise.

**Comment by balacij** [comment]
https://github.com/JacquesCarette/Drasil/issues/1137#issuecomment-625992247 says that the original problem is 'resolved' but I also think the leftovers is resolved (?). To my knowledge, Drasil should never allow symbols to have non-unqiue `Symbol`s (w.r.t. a particular `ChunkDB`).

**Comment by JacquesCarette** [comment]
Actually, ruling out non-unique `Symbol` s, even wrt a particular `ChunkDB`, is probably not what we want. People really do overload notation in all sorts of ways. Agda uses it nicely so that `suc` for example can be for both Nat and Fin and is type-disambiguated at use site.

We do need something like a "rendering context" within which a `Symbol` has to refer to a unique entity. What 'unique' means is actually context-sensitive, i.e. it will depend on the rules of the ambient PL that we're generating. I think the ones we use are fairly straightforward, but I think even Scala allows for some funky stuff.
